src/lib/idmap/sss_idmap.h

```diff
@@ -956,6 +961,136 @@ enum idmap_error_code sss_idmap_smb_sid_to_bin_sid(struct sss_idmap_ctx *ctx,
+ *                        use the realted ranges for mapping.
+ *
+ * @return
+ *  - #IDMAP_OUT_OF_MEMORY: Insufficient memory to store the data in the idmap
+ *                          context
+ *  - #IDMAP_SID_INVALID:   Invalid SID provided
+ *  - #IDMAP_NO_DOMAIN:     No domain domain name given
+ *  - #IDMAP_COLLISION:     New domain collides with existing one
+ */
+enum idmap_error_code sss_idmap_add_gen_domain_ex(struct sss_idmap_ctx *ctx,
+                                                  const char *domain_name,
+                                                  const char *domain_id,
+                                                  struct sss_idmap_range *range,
+                                                  const char *range_id,
+                                                  idmap_offset_func *offset_func,
+                                                  idmap_rev_offset_func *rev_offset_func,
+                                                  void *offset_func_pvt,
+                                                  uint32_t shift,
+                                                  bool external_mapping);
+
+/**
+ * @brief Calculate offset from string containing only numbers
+ */
+enum idmap_error_code sss_idmap_offset_identity(void *pvt, uint32_t range_size,
+                                                const char *input,
+                                                long long *offset);
+
+/**
+ * @brief Reverse of sss_idmap_offset_identity, return a string containg only
+ * numbers representing the given offset
+ */
+enum idmap_error_code sss_idmap_rev_offset_identity(struct sss_idmap_ctx *ctx,
+                                                    void *pvt, uint32_t id,
+                                                    char **_out);
+
+/**
+ * @brief Calculate offset from string with the help of murmurhash3
+ */
+enum idmap_error_code sss_idmap_offset_murmurhash3(void *pvt,
+                                                   uint32_t range_size,
+                                                   const char *input,
+                                                   long long *offset);
+
+/**
+ * Structure for private data for offset_murmurhash3. If not given 0xdeadbeef
+ * will be used as seed.
+ */
+struct sss_idmap_offset_murmurhash3_data {
+    uint32_t seed;
+};
+
+/**
+ * @brief Translate some input to a unix UID or GID
+ *
+ * @param[in] ctx       Idmap context
+ * @param[in] domain_id Zero-terminated string with the domain ID of a known
+ *                      domain
+ * @param[in] input     Zero-terminated string which should be translated into
+ *                      an offset to calculate the unix UID or GID
+ * @param[out] id       Returned unix UID or GID
+ *
+ * @return
+ *  - #IDMAP_NO_DOMAIN:     No domain with domain_id found in ctx
+ *  - #IDMAP_EXTERNAL:      external source is authoritative for mapping
+ */
+enum idmap_error_code sss_idmap_gen_to_unix(struct sss_idmap_ctx *ctx,
```

Comment:
Here and in `sss_idmap_unix_to_gen()` `ctx` could be a ptr to const?

Reply:
I would prefer to keep it as it is because currently `const` is never used for `struct sss_idmap_ctx`. Maybe with a general API redesign (see other comment) it can be added.